### DISCUSSION TITLE
Handling Queue Processing in a NestJS App Deployed on Vercel

### DISCUSSION THREAD

**Discussion by vksazonov196** [community/community#136271] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

I have a NestJS app that handles receiving and writing messages on Telegram. To manage these messages efficiently, I integrated BullMQ to distribute them into different queues. Everything works perfectly when running the app locally; the messages are successfully added to the queue and processed as expected.

However, once deployed to Vercel, a problem arises. When I send a message, it gets added to the queue, but processing halts until I start the app locally again. I'm not sure, but in my opinion this issue is likely due to Vercel's serverless architecture, which doesn't support long-running processes like queue workers. As a result, the queue remains idle until the app is run locally, where it can process the tasks.
Has anyone encountered a similar issue? Or can anyone recommend a better service to deploy my app that supports long-running tasks?

**Comment by georgezouq** [comment]
Vercel is severless, when you queue added, it couldn't receive. 

Any solution for this problem?

**Comment by PakistanAiFrontierCorps** [comment]
You’re right, the issue comes from Vercel’s serverless model. Vercel functions are short lived and shut down once they return a response, so they can’t run a BullMq worker. That’s the reason your jobs sit in Redis until you start the worker locally.

If you want to stick with BullMQ, you need to run the worker on a service that supports long-running processes. Typical options are:

1. Railway, Render, Fly.io, or Heroku → run the worker in a container that stays alive.

2. VPS providers → run both API and worker together.

3. Hybrid setup → keep your API on Vercel, but deploy only the worker on Railway/Render/Fly.io and connect them to the same Redis

Brutal truth: you will not get BullMQ workers running on Vercel alone. Use Vercel just for API/frontend, and move your background workers to a service that supports persistent processes.
